diff -Naur a/cmd/configure.ac b/cmd/configure.ac
--- a/cmd/configure.ac	2026-03-24 15:10:56.936399727 +0300
+++ b/cmd/configure.ac	2026-03-24 15:18:54.257926728 +0300
@@ -249,9 +249,6 @@
 dnl FIXME: get this via something like pkgconf once it is defined there
 dnl FIXME: Use PKG_CHECK_VAR when we have dropped Trusty (14.04)
 AC_ARG_VAR([SYSTEMD_PREFIX], [value for systemd prefix (overriding pkg-config)])
-if test -z "${SYSTEMD_PREFIX}"; then
-  SYSTEMD_PREFIX="$($PKG_CONFIG --variable=prefix systemd)"
-fi
 if test -n "${SYSTEMD_PREFIX}"; then
   SYSTEMD_SYSTEM_ENV_GENERATOR_DIR="${SYSTEMD_PREFIX}/lib/systemd/system-environment-generators"
 else
diff -Naur a/cmd/libsnap-confine-private/utils.c b/cmd/libsnap-confine-private/utils.c
--- a/cmd/libsnap-confine-private/utils.c	2026-03-24 15:10:56.941400933 +0300
+++ b/cmd/libsnap-confine-private/utils.c	2026-03-24 15:19:31.296128057 +0300
@@ -217,7 +217,7 @@
 
 bool sc_is_expected_path(const char *path) {
     const char *expected_path_re =
-        "^((/var/lib/snapd)?/snap/(snapd|core)/x?[0-9]+/usr/lib|/usr/lib(exec)?)/snapd/snap-confine$";
+        "^((/var/lib/snapd)?/snap/(snapd|core)/x?[0-9]+/usr/lib|(/usr|@out@)/lib(exec)?)/snapd/snap-confine(-unwrapped)?$";
     regex_t re;
     if (regcomp(&re, expected_path_re, REG_EXTENDED | REG_NOSUB) != 0)
         die("can not compile regex %s", expected_path_re);
diff -Naur a/cmd/snap-confine/mount-support.c b/cmd/snap-confine/mount-support.c
--- a/cmd/snap-confine/mount-support.c	2026-03-24 15:10:56.945898491 +0300
+++ b/cmd/snap-confine/mount-support.c	2026-03-24 15:20:31.944939273 +0300
@@ -921,7 +921,7 @@
             {.path = "/run"},            // to get /run with sockets and what not
             {.path = "/lib/modules", .is_optional = true},   // access to the modules of the running kernel
             {.path = "/lib/firmware", .is_optional = true},  // access to the firmware of the running kernel
-            {.path = "/usr/src"},                            // FIXME: move to SecurityMounts in system-trace interface
+            {.path = "/usr/src", .is_optional = true},                            // FIXME: move to SecurityMounts in system-trace interface
             {.path = "/var/log"},                            // FIXME: move to SecurityMounts in log-observe interface
 #ifdef MERGED_USR
             {.path = "/run/media",
diff -Naur a/cmd/snap-confine/seccomp-support.c b/cmd/snap-confine/seccomp-support.c
--- a/cmd/snap-confine/seccomp-support.c	2026-03-24 15:10:56.946549418 +0300
+++ b/cmd/snap-confine/seccomp-support.c	2026-03-24 15:21:17.430273967 +0300
@@ -80,10 +80,6 @@
     if (stat_buf.st_uid != 0 || stat_buf.st_gid != 0) {
         die("%s not root-owned %i:%i", path, stat_buf.st_uid, stat_buf.st_gid);
     }
-
-    if (stat_buf.st_mode & S_IWOTH) {
-        die("%s has 'other' write %o", path, stat_buf.st_mode);
-    }
 }
 
 static void validate_bpfpath_is_safe(const char *path) {
diff -Naur a/cmd/snap-confine/snap-confine.c b/cmd/snap-confine/snap-confine.c
--- a/cmd/snap-confine/snap-confine.c	2026-03-24 15:10:56.947333084 +0300
+++ b/cmd/snap-confine/snap-confine.c	2026-03-24 15:21:52.767911251 +0300
@@ -525,7 +525,8 @@
          * one, which definitely doesn't run in a snap-specific namespace, has a
          * predictable PID and is long lived.
          */
-        sc_reassociate_with_pid1_mount_ns();
+        // sc_reassociate_with_pid1_mount_ns();
+        // NIX PATCH: Don't try to escape the FHS environment
         // Do global initialization:
         int global_lock_fd = sc_lock_global();
         // Ensure that "/" or "/snap" is mounted with the
diff -Naur a/dirs/dirs.go b/dirs/dirs.go
--- a/dirs/dirs.go	2026-03-24 15:10:56.993347060 +0300
+++ b/dirs/dirs.go	2026-03-24 15:24:33.255169426 +0300
@@ -406,7 +406,7 @@
 // SnapServicesDirUnder returns the path to the systemd services
 // conf dir under rootdir.
 func SnapServicesDirUnder(rootdir string) string {
-	return filepath.Join(rootdir, "/etc/systemd/system")
+	return filepath.Join(rootdir, "/var/lib/snapd/nix-systemd-system")
 }
 
 func SnapRuntimeServicesDirUnder(rootdir string) string {
@@ -645,7 +645,7 @@
 	SnapSystemdDir = filepath.Join(rootdir, "/etc/systemd")
 	SnapSystemdRunDir = filepath.Join(rootdir, "/run/systemd")
 
-	SnapDBusSystemPolicyDir = filepath.Join(rootdir, "/etc/dbus-1/system.d")
+	SnapDBusSystemPolicyDir = filepath.Join(rootdir, "/var/lib/snapd/nix-dbus-system")
 	SnapDBusSessionPolicyDir = filepath.Join(rootdir, "/etc/dbus-1/session.d")
 	// Use 'dbus-1/services' and `dbus-1/system-services' to mirror
 	// '/usr/share/dbus-1' hierarchy.
@@ -657,7 +657,7 @@
 
 	CloudInstanceDataFile = filepath.Join(rootdir, "/run/cloud-init/instance-data.json")
 
-	SnapUdevRulesDir = filepath.Join(rootdir, "/etc/udev/rules.d")
+	SnapUdevRulesDir = filepath.Join(rootdir, "/var/lib/snapd/nix-udev-rules")
 
 	SnapKModModulesDir = filepath.Join(rootdir, "/etc/modules-load.d/")
 	SnapKModModprobeDir = filepath.Join(rootdir, "/etc/modprobe.d/")
@@ -668,7 +668,7 @@
 	LocaleDir = filepath.Join(rootdir, "/usr/share/locale")
 	ClassicDir = filepath.Join(rootdir, "/writable/classic")
 
-	DistroLibExecDir = filepath.Join(rootdir, DefaultDistroLibexecDir)
+	DistroLibExecDir = filepath.Join(rootdir, "@out@/libexec/snapd")
 	if _, err := os.Stat(DistroLibExecDir); errors.Is(err, fs.ErrNotExist) {
 		// the default /usr/lib/snapd does not exist, but maybe we have the
 		// alternative dir /usr/libexec/snapd
diff -Naur a/interfaces/system_key.go b/interfaces/system_key.go
--- a/interfaces/system_key.go	2026-03-24 15:10:57.038148712 +0300
+++ b/interfaces/system_key.go	2026-03-24 15:25:26.289689084 +0300
@@ -122,7 +122,7 @@
 	sk := &systemKey{
 		Version: systemKeyVersion,
 	}
-	snapdPath, err := snapdtool.InternalToolPath("snapd")
+	snapdPath, err := snapdtool.InternalToolPath("snapd-unwrapped")
 	if err != nil {
 		return nil, err
 	}
@@ -291,7 +291,7 @@
 	if mockedSystemKey == nil {
 		if exe, err := os.Readlink("/proc/self/exe"); err == nil {
 			// detect running local local builds
-			if !strings.HasPrefix(exe, "/usr") && !strings.HasPrefix(exe, dirs.SnapMountDir) {
+			if !strings.HasPrefix(exe, "/usr") && !strings.HasPrefix(exe, dirs.SnapMountDir) && !strings.HasPrefix(exe, "@out@") {
 				logger.Noticef("running from non-installed location %s: ignoring system-key", exe)
 				return false, mySystemKey, ErrSystemKeyVersion
 			}
diff -Naur a/snap/info.go b/snap/info.go
--- a/snap/info.go	2026-03-24 15:10:57.128472023 +0300
+++ b/snap/info.go	2026-03-24 15:26:08.825705571 +0300
@@ -1527,9 +1527,9 @@
 		command = " " + command
 	}
 	if app.Name == app.Snap.SnapName() {
-		return fmt.Sprintf("/usr/bin/snap run%s %s", command, app.Snap.InstanceName())
+		return fmt.Sprintf("@out@/bin/snap run%s %s", command, app.Snap.InstanceName())
 	}
-	return fmt.Sprintf("/usr/bin/snap run%s %s.%s", command, app.Snap.InstanceName(), app.Name)
+	return fmt.Sprintf("@out@/bin/snap run%s %s.%s", command, app.Snap.InstanceName(), app.Name)
 }
 
 // LauncherCommand returns the launcher command line to use when invoking the
diff -Naur a/systemd/systemd.go b/systemd/systemd.go
--- a/systemd/systemd.go	2026-03-24 15:10:57.142303296 +0300
+++ b/systemd/systemd.go	2026-03-24 15:29:15.847494623 +0300
@@ -618,6 +618,14 @@
 	if len(serviceNames) == 0 {
 		return nil
 	}
+	for _, serviceName := range serviceNames {
+		servicePath := filepath.Join(dirs.SnapServicesDir, serviceName)
+		serviceRuntimePath := filepath.Join(dirs.SnapRuntimeServicesDir, serviceName)
+		os.Remove(serviceRuntimePath)
+		if err := os.Symlink(servicePath, serviceRuntimePath); err != nil {
+			return err
+		}
+	}
 	var args []string
 	if s.rootDir != "" {
 		// passing root already implies no reload
@@ -625,6 +633,7 @@
 	} else {
 		args = append(args, "--no-reload")
 	}
+	args = append(args, "--runtime")
 	args = append(args, "enable")
 	args = append(args, serviceNames...)
 	_, err := s.systemctl(args...)
@@ -645,6 +654,14 @@
 	if len(serviceNames) == 0 {
 		return nil
 	}
+	for _, serviceName := range serviceNames {
+		servicePath := filepath.Join(dirs.SnapServicesDir, serviceName)
+		serviceRuntimePath := filepath.Join(dirs.SnapRuntimeServicesDir, serviceName)
+		os.Remove(serviceRuntimePath)
+		if err := os.Symlink(servicePath, serviceRuntimePath); err != nil {
+			return err
+		}
+	}
 	var args []string
 	if s.rootDir != "" {
 		// passing root already implies no reload
@@ -652,6 +669,7 @@
 	} else {
 		args = append(args, "--no-reload")
 	}
+	args = append(args, "--runtime")
 	args = append(args, "disable")
 	args = append(args, serviceNames...)
 	_, err := s.systemctl(args...)
diff -Naur a/usersession/userd/launcher.go b/usersession/userd/launcher.go
--- a/usersession/userd/launcher.go	2026-03-24 15:10:57.365637371 +0300
+++ b/usersession/userd/launcher.go	2026-03-24 15:30:04.819541588 +0300
@@ -206,7 +206,7 @@
 	// this code must not add directories from the snap
 	// to XDG_DATA_DIRS and similar, see
 	// https://ubuntu.com/security/CVE-2020-11934
-	if err := exec.Command("xdg-open", addr).Run(); err != nil {
+	if err := exec.Command("@out@/libexec/xdg-open", addr).Run(); err != nil {
 		return dbus.MakeFailedError(fmt.Errorf("cannot open supplied URL"))
 	}
 
@@ -292,7 +292,7 @@
 		return dbus.MakeFailedError(fmt.Errorf("permission denied"))
 	}
 
-	if err = exec.Command("xdg-open", filename).Run(); err != nil {
+	if err = exec.Command("@out@/libexec/xdg-open", filename).Run(); err != nil {
 		return dbus.MakeFailedError(fmt.Errorf("cannot open supplied URL"))
 	}
 
diff -Naur a/wrappers/binaries.go b/wrappers/binaries.go
--- a/wrappers/binaries.go	2026-03-24 15:10:57.366605116 +0300
+++ b/wrappers/binaries.go	2026-03-24 15:30:39.827570322 +0300
@@ -220,7 +220,7 @@
 		}
 
 		appBase := filepath.Base(app.WrapperPath())
-		binariesContent[appBase] = &osutil.SymlinkFileState{Target: "/usr/bin/snap"}
+		binariesContent[appBase] = &osutil.SymlinkFileState{Target: "@out@/bin/snap"}
 
 		if completionVariant != noCompletion && app.Completer != "" {
 			completersContent[appBase] = &osutil.SymlinkFileState{Target: completeSh}
diff -Naur a/wrappers/internal/service_unit_gen.go b/wrappers/internal/service_unit_gen.go
--- a/wrappers/internal/service_unit_gen.go	2026-03-24 15:10:57.367927830 +0300
+++ b/wrappers/internal/service_unit_gen.go	2026-03-24 15:31:12.783111246 +0300
@@ -187,6 +187,7 @@
 TimeoutStartSec={{.StartTimeout}}
 {{- end}}
 Type={{.App.Daemon}}
+NotifyAccess=all
 {{- if .Remain}}
 RemainAfterExit={{.Remain}}
 {{- end}}
